src/smolvm/vm.py

```
@@ -545,6 +547,166 @@ def _copy_with_reflink(source_path: Path, target_path: Path) -> None:
+        resize2fs = which("resize2fs")
+        if e2fsck is None or resize2fs is None:
+            raise SmolVMError(
+                f"e2fsck and resize2fs are needed to grow the disk for sandbox '{vm_id}'; "
+                f"install e2fsprogs, or run '{self._resize_recovery(vm_id)}'."
+            )
+        self._run_resize_tool([str(e2fsck), "-fy", str(disk_path)], "e2fsck", vm_id)
+        self._run_resize_tool([str(resize2fs), str(disk_path)], "resize2fs", vm_id)
+
+    def _run_resize_tool(self, command: list[str], tool_name: str, vm_id: str) -> None:
+        """Run one resize helper and convert failures to SmolVMError."""
+        result = subprocess.run(command, capture_output=True, text=True, check=False)
```

Ruff (0.15.15)

[error] `subprocess` call: check for execution of untrusted input

(S603)